### PR TITLE
commit-graph: Add a way to write commit-graph files

### PR DESCRIPTION
**Pull request by lhchavez**
This change adds the git_commit_graph_writer_* functions to allow to
write and create `commit-graph` files from `.idx`/`.pack` files or
`git_revwalk`s.

Part of: #5757

### CODE REVIEW

File: src/commit_graph.c [hunk 1]
```diff
@@ -36,6 +44,63 @@ struct git_commit_graph_chunk {
...
+	git_array_clear(p->parents);
+	git_array_clear(p->parent_indices);
+	git__free(p);
+}
+
+static struct packed_commit *packed_commit_new(git_commit *commit)
+{
+	unsigned int i, parentcount = git_commit_parentcount(commit);
+	struct packed_commit *p
+			= (struct packed_commit *)git__calloc(1, sizeof(struct packed_commit));
```
Comment (ethomson) [inline src/commit_graph.c:73]: We don't cast allocations

File: src/commit_graph.c [hunk 2]
```diff
@@ -547,3 +611,583 @@ void git_commit_graph_file_free(git_commit_graph_file *file)
...
+
+int git_commit_graph_writer_add_index_file(
+		git_commit_graph_writer *w,
+		git_repository *repo,
+		const char *idx_path)
+{
+	int error;
+	struct git_pack_file *p = NULL;
+	struct object_entry_cb_state state = {
+			.repo = repo,
```
Comment (ethomson) [inline src/commit_graph.c:704]: I know we've gone back and forth a bit on our minimum C standard but we haven't been using the struct dot initializer syntax and I'd like to be more thoughtful about whether we start or not.  Can we just use regular assignments here for now.

File: src/commit_graph.c [hunk 3]
```diff
@@ -547,3 +611,583 @@ void git_commit_graph_file_free(git_commit_graph_file *file)
...
+struct commit_graph_write_hash_context {
+	commit_graph_write_cb write_cb;
+	void *cb_data;
+	git_hash_ctx *ctx;
+};
+
+static int commit_graph_write_hash(const char *buf, size_t size, void *data)
+{
+	struct commit_graph_write_hash_context *ctx
+			= (struct commit_graph_write_hash_context *)data;
```
Comment (ethomson) [inline src/commit_graph.c:945]: We don't cast allocs...

File: src/commit_graph.c [hunk 4]
```diff
@@ -547,3 +611,583 @@ void git_commit_graph_file_free(git_commit_graph_file *file)
...
+}
+
+static int
+commit_graph_write(git_commit_graph_writer *w, commit_graph_write_cb write_cb, void *cb_data)
+{
+	int error = 0;
+	size_t i;
+	struct packed_commit *packed_commit;
+	struct git_commit_graph_header hdr = {
+			.signature = htonl(COMMIT_GRAPH_SIGNATURE),
```
Comment (ethomson) [inline src/commit_graph.c:967]: Let's not use the dot initializer syntax.

File: src/commit_graph.c [hunk 5]
```diff
@@ -547,3 +611,583 @@ void git_commit_graph_file_free(git_commit_graph_file *file)
...
+	uint32_t oid_fanout_count;
+	uint32_t extra_edge_list_count;
+	uint32_t oid_fanout[256];
+	off64_t offset;
+	git_buf oid_lookup = GIT_BUF_INIT,
+		commit_data = GIT_BUF_INIT, extra_edge_list = GIT_BUF_INIT;
+	git_oid cgraph_checksum = {{0}};
+	git_hash_ctx ctx;
+	struct commit_graph_write_hash_context hash_cb_data = {
+			.write_cb = write_cb,
```
Comment (ethomson) [inline src/commit_graph.c:982]: Dot initializer syntax

File: src/commit_graph.c [hunk 6]
```diff
@@ -547,3 +611,583 @@ void git_commit_graph_file_free(git_commit_graph_file *file)
...
+	/* Fill the Commit Data and Extra Edge List tables. */
+	extra_edge_list_count = 0;
+	git_vector_foreach (&w->commits, i, packed_commit) {
+		uint64_t commit_time;
+		uint32_t generation;
+		uint32_t word;
+		unsigned int parentcount = (unsigned int)git_array_size(packed_commit->parents);
+
+		error
+				= git_buf_put(&commit_data,
```
Comment (ethomson) [inline src/commit_graph.c:1028]: So much whitespace

File: src/commit_graph.c [hunk 7]
```diff
@@ -547,3 +611,583 @@ void git_commit_graph_file_free(git_commit_graph_file *file)
...
+	if (error < 0)
+		goto cleanup;
+	error = write_cb(git_buf_cstr(&oid_lookup), git_buf_len(&oid_lookup), cb_data);
+	if (error < 0)
+		goto cleanup;
+	error = write_cb(git_buf_cstr(&commit_data), git_buf_len(&commit_data), cb_data);
+	if (error < 0)
+		goto cleanup;
+	error
+			= write_cb(git_buf_cstr(&extra_edge_list),
```
Comment (ethomson) [inline src/commit_graph.c:1126]: so much whitespace here too

File: include/git2/sys/commit_graph.h [hunk 8]
```diff
@@ -40,6 +40,136 @@ GIT_EXTERN(int) git_commit_graph_open(git_commit_graph **cgraph_out, const char
...
+/**
+ * Write a `commit-graph` file to a file.
+ *
+ * @param w The writer
+ * @param opts Pointer to git_commit_graph_writer_options struct.
+ * @return 0 or an error code
+ */
+GIT_EXTERN(int) git_commit_graph_writer_commit(
+		git_commit_graph_writer *w,
+		git_commit_graph_writer_options *opts);
```
Comment (ethomson) [inline include/git2/sys/commit_graph.h:158]: I would have expected `git_commit_graph_writer_new` to take the options and store them...  it's a little unclear how these interact, it looks like all the callers ignore them at the moment?

**Review by ethomson — COMMENTED** [review]
Thanks again for your work on commit graphs, this is really exciting stuff.  A few minor nits and a few questions.  One additional question - are you using some tool to format your code?  Your formatting seems consistent with itself but not particularly consistent with the rest of the project.  Curious if this is your style, a tool, etc...

**Comment by lhchavez** [comment]
> Thanks again for your work on commit graphs, this is really exciting stuff. A few minor nits and a few questions. One additional question - are you using some tool to format your code? Your formatting seems consistent with itself but not particularly consistent with the rest of the project. Curious if this is your style, a tool, etc...

yes! i am using clang-format with settings that _tried_ to be as consistent as possible with the rest of the codebase (but apparently didn't meet the goal u_u).

File: include/git2/sys/commit_graph.h (hunk 8 above)
Comment (lhchavez) [reply include/git2/sys/commit_graph.h:158]: yup, it's mostly there to avoid breaking API/ABI in the future.

File: src/commit_graph.c (hunk 3 above)
Comment (lhchavez) [reply src/commit_graph.c:945]: this one is not an alloc, is it still not acceptable to cast from a `void *`?

File: include/git2/sys/commit_graph.h (hunk 8 above)
Comment (lhchavez) [reply include/git2/sys/commit_graph.h:158]: Added a TODO to support them in a later commit to make that clear.

File: src/commit_graph.c (hunk 6 above)
Comment (lhchavez) [reply src/commit_graph.c:1028]: Nothing that a little `PenaltyBreakAssignment: 1000` can't fix.

File: src/commit_graph.c (hunk 1 above)
Comment (lhchavez) [reply src/commit_graph.c]: No more casts!

File: src/commit_graph.c (hunk 2 above)
Comment (lhchavez) [reply src/commit_graph.c]: No more dot inializer syntax

**Comment by lhchavez** [comment]
> > Thanks again for your work on commit graphs, this is really exciting stuff. A few minor nits and a few questions. One additional question - are you using some tool to format your code? Your formatting seems consistent with itself but not particularly consistent with the rest of the project. Curious if this is your style, a tool, etc...
> 
> yes! i am using clang-format with settings that _tried_ to be as consistent as possible with the rest of the codebase (but apparently didn't meet the goal u_u).

these are the current settings i have:

```
# .clang-format
---
AlignAfterOpenBracket: AlwaysBreak
AlignConsecutiveAssignments: false
AlignTrailingComments: true
AlignOperands: false
AllowAllParametersOfDeclarationOnNextLine: false
AllowShortBlocksOnASingleLine: false
AllowShortFunctionsOnASingleLine: false
AllowShortIfStatementsOnASingleLine: false
BasedOnStyle: WebKit
BinPackArguments: false
BinPackParameters: false
BreakBeforeBraces: Linux
ColumnLimit: 100
ContinuationIndentWidth: 16
Cpp11BracedListStyle: true
ForEachMacros:
- git_vector_foreach
- git_vector_rforeach
- git_array_foreach
IncludeCategories:
- Priority: 1
  Regex: ^<
- Priority: 2
  Regex: ^[<"]git2/
- Priority: 3
  Regex: ^"
IndentCaseLabels: false
IndentWidth: 8
MaxEmptyLinesToKeep: 2
PointerAlignment: Right
SortIncludes: false
SpacesBeforeTrailingComments: 2
TypenameMacros:
- GIT_EXTERN
UseTab: Always
PenaltyBreakAssignment: 1000
...
```

**Comment by ethomson** [comment]
Sorry it's taken so long to get back to this, #5974 took a lot more time than I expected.  😢 

One question about the struct initializer.  Otherwise, I'm going to take a whack at tweaking clang-format to see if we can get it a little closer to our style.

Then I'm looking forward to merging this and cutting 1.2.0 with all your improvements.  Thanks for all the hard work - and your patience in getting them reviewed.

**Comment by ethomson** [comment]
I did a few formatting fixes because I had it beaten into my head that we should have binary operators at the end of a line not the beginning.  I apologize for this compulsion, but there it is.

Manually merged.  Thanks for all this work!

**Pull request merged**